Mathlib/Util/CountHeartbeats.lean

```diff
@@ -24,6 +24,59 @@ open Lean Elab Command Meta
 namespace Mathlib.CountHeartbeats
 
 
+
+open Tactic
+
+/--
+Run a tactic, optionally restoring the original state, and report just the number of heartbeats.
+-/
+def runTacForHeartbeats (tac : TSyntax `Lean.Parser.Tactic.tacticSeq) (revert : Bool := true) :
+    TacticM Nat := do
+  let start ← IO.getNumHeartbeats
+  let s ← saveState
+  evalTactic tac
+  if revert then restoreState s
+  return (← IO.getNumHeartbeats) - start
+
+/--
+Given a `List Nat`, return the minimum, maximum, and standard deviation.
+-/
+def variation (counts : List Nat) : List Nat :=
+  let min := counts.minimum?.getD 0
+  let max := counts.maximum?.getD 0
+  let μ := counts.foldl (· + ·) 0 / counts.length
+  -- We jump through some hoops here to get access to Float.sqrt, to avoid imports.
+  let stddev := (Float.sqrt <| UInt64.toFloat <| UInt64.ofNat
+    -- `i - μ + (μ - i)` really computes `|i - μ|` with non-truncated subtraction
+    (((counts.map fun i => (i - μ + (μ - i))^2).foldl (· + ·) 0) / counts.length)).toUInt64.toNat
+  [min, max, stddev]
```

Comment:
There's a bug here where the standard deviation is inaccurate due the mean being rounded to an integer. May as well do all the calculations as floats, which makes the logic clearer I think.
```suggestion
def variation (counts : List Nat) : List Nat :=
  let min := counts.minimum?.getD 0
  let max := counts.maximum?.getD 0
  let toFloat (n : Nat) := n.toUInt64.toFloat
  let toNat (f : Float) := f.toUInt64.toNat
  let counts' := counts.map toFloat
  let μ : Float := counts'.foldl (· + ·) 0 / toFloat counts.length
  let stddev : Float := Float.sqrt <|
    ((counts'.map fun i => (i - μ)^2).foldl (· + ·) 0) / toFloat counts.length
  [min, max, toNat stddev]
```